taichi/backends/opengl/opengl_device.h

```diff
@@ -0,0 +1,246 @@
+  // index_width = 4 -> uint32 index
+  // index_width = 2 -> uint16 index
+  void index_buffer(DevicePtr ptr, size_t index_width) override;
+
+  const std::unordered_map<uint32_t, GLuint> &binding_map() {
+    return binding_map_;
+  }
+
+ private:
+  std::unordered_map<uint32_t, GLuint> binding_map_;
+};
+
+class GLPipeline : public Pipeline {
+ public:
+  GLPipeline(const PipelineSourceDesc &desc, std::string name);
+  ~GLPipeline() override;
+
+  ResourceBinder *resource_binder() override;
+
+  GLuint get_program() {
+    return program_id_;
+  }
+
+ private:
+  GLuint program_id_;
+  GLResourceBinder binder_;
+};
+
+class GLCommandList : public CommandList {
+ public:
+  ~GLCommandList() override;
+
+  void bind_pipeline(Pipeline *p) override;
+  void bind_resources(ResourceBinder *binder) override;
+  void bind_resources(ResourceBinder *binder,
+                      ResourceBinder::Bindings *bindings) override;
+  void buffer_barrier(DevicePtr ptr, size_t size) override;
+  void buffer_barrier(DeviceAllocation alloc) override;
+  void memory_barrier() override;
+  void buffer_copy(DevicePtr dst, DevicePtr src, size_t size) override;
+  void buffer_fill(DevicePtr ptr, size_t size, uint32_t data) override;
+  void dispatch(uint32_t x, uint32_t y = 1, uint32_t z = 1) override;
+
+  // These are not implemented in compute only device
+  void begin_renderpass(int x0,
+                        int y0,
+                        int x1,
+                        int y1,
+                        uint32_t num_color_attachments,
+                        DeviceAllocation *color_attachments,
+                        bool *color_clear,
+                        std::vector<float> *clear_colors,
+                        DeviceAllocation *depth_attachment,
+                        bool depth_clear) override;
+  void end_renderpass() override;
+  void draw(uint32_t num_verticies, uint32_t start_vertex = 0) override;
+  void clear_color(float r, float g, float b, float a) override;
+  void set_line_width(float width) override;
+  void draw_indexed(uint32_t num_indicies,
+                    uint32_t start_vertex = 0,
+                    uint32_t start_index = 0) override;
+  void image_transition(DeviceAllocation img,
+                        ImageLayout old_layout,
+                        ImageLayout new_layout) override;
+  void buffer_to_image(DeviceAllocation dst_img,
+                       DevicePtr src_buf,
+                       ImageLayout img_layout,
+                       const BufferImageCopyParams &params) override;
+  void image_to_buffer(DevicePtr dst_buf,
+                       DeviceAllocation src_img,
+                       ImageLayout img_layout,
+                       const BufferImageCopyParams &params) override;
+
+  // GL only stuff
+  void run_commands();
+
+ private:
+  struct Cmd {
+    virtual void execute() {
+    }
+  };
+
+  struct CmdBindPipeline : public Cmd {
+    GLuint program;
```

Comment:
nit: Init these member variables?

Reply:
I thought since these are private member are only ever created through functions within this class, the overhead of initialization could be saved

Reply:
The initialization here is mostly to make the new code looks consistent :-) (And I don't think there will be any discernible difference here..)